src/rustc_lexer/src/lib.rs

```diff
@@ -0,0 +1,311 @@
+                    self.hex_digits()
+                },
+                '0'..='9' | '_' | '.' | 'e' | 'E' => {
+                    self.decimal_digits();
+                    true
+                }
+                _ => {
+                    // just a 0
+                    return Int { empty_int: false };
+                }
+            };
+            if !has_digits {
+                return Int { empty_int: true };
+            }
+        } else {
+            self.decimal_digits();
+        };
+
+        match self.nth_char(0) {
+            // Don't be greedy if this is actually an
+            // integer literal followed by field/method access or a range pattern
+            // (`0..2` and `12.foo()`)
+            '.' if self.nth_char(1) != '.' && !char_props::xid_start(self.nth_char(1)) => {
+                // might have stuff after the ., and if it does, it needs to start
+                // with a number
+                self.bump();
+                let mut empty_exponent = false;
+                if self.nth_char(0).is_digit(10) {
+                    self.decimal_digits();
+                    match self.nth_char(0) {
+                        'e' | 'E' => {
+                            self.bump();
+                            empty_exponent = self.float_exponent().is_err()
+                        }
+                        _ => (),
+                    }
+                }
+                Float { empty_exponent }
+            }
+            'e' | 'E' => {
+                self.bump();
+                let empty_exponent = self.float_exponent().is_err();
+                Float { empty_exponent }
+            }
+            _ => Int { empty_int: false },
+        }
+    }
+
+    fn decimal_digits(&mut self) -> bool {
+        let mut has_digits = false;
+        loop {
+            match self.nth_char(0) {
+                '_' => {
+                    self.bump();
+                }
+                '0' ..= '9' => {
+                    has_digits = true;
+                    self.bump();
+                },
+                _ => break,
+            }
+        }
+        has_digits
+    }
+
+    fn hex_digits(&mut self) -> bool {
+        let mut has_digits = false;
+        loop {
+            match self.nth_char(0) {
+                '_' => {
+                    self.bump();
+                }
+                '0' ..= '9' | 'a' ..= 'f' | 'A' ..= 'F' => {
+                    has_digits = true;
+                    self.bump();
+                },
+                _ => break,
+            }
+        }
+        has_digits
+    }
+
+    fn float_exponent(&mut self) -> Result<(), ()> {
+        assert!(self.prev() == 'e' || self.prev() == 'E');
+        if self.nth_char(0) == '-' || self.nth_char(0) == '+' {
+            self.bump();
+        }
+        if self.decimal_digits() {
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+
+    fn literal_suffix(&mut self) -> bool {
```

Comment:
I would use a convention here that I've seen elsewhere, which is to name methods that *may* consume (and which return `true` if they *did* so), with an `eat_` prefix.
This makes code calling them look like `if self.eat_literal_suffix() {...}`, which is IMO more readable.
cc @petrochenkov

Reply:
Great idea, thanks!